### PR TITLE
breaking: JSON unescape

### PR DESCRIPTION
**Pull request by xackus**
Unescape JSON strings in the parser.
Looking at the comments in the code suggests it was planned but not done.

Even if we want it as an option, I think escaping should be the default.
I can add the option if there is need.

### CODE REVIEW

**Comment by xackus** [comment]
Maybe `unescapeStringAlloc` should stay as an internal function. I didn't add bounds checks, because it operated on already validated strings in the parser. Then i thought I could make it available for general use and forgot about it.

**Comment by xackus** [comment]
I also just came up with this idea: the parser could count how many bytes will be needed to store the unescaped string and store it in the token. Then you could allocate the correct amount of memory without going through the string another time.

File: lib/std/json.zig [hunk 1]
```diff
@@ -1270,6 +1270,47 @@ pub const Parser = struct {
     }
 };
 
+/// Unescape a JSON string
+/// Optimized for arena allocators, uses Allocator.shrink
+pub fn unescapeStringAlloc(alloc: *Allocator, input: []const u8) ![]u8 {
+    var output = try alloc.alloc(u8, input.len);
```
Comment (daurnimator) [inline lib/std/json.zig]: This should be `const`

File: lib/std/json.zig [hunk 2]
```diff
@@ -1379,3 +1420,34 @@ test "parsing empty string gives appropriate error" {
...
+        \\{
+        \\  "backslash": "\\",
+        \\  "forwardslash": "\/",
+        \\  "newline": "\n",
+        \\  "carriagereturn": "\r",
+        \\  "tab": "\t",
+        \\  "formfeed": "\f",
+        \\  "backspace": "\b",
+        \\  "doublequote": "\"",
+        \\  "unicode": "\u0105"
```
Comment (daurnimator) [inline lib/std/json.zig]: Could you add some tests with invalid escapes? e.g. `\x`,  an underlong unicode sequence `\u2abc`

Comment (xackus) [reply lib/std/json.zig]: Invalid escapes are caught earlier in the parser and already tested in json/test.zig

**Comment by xackus** [comment]
Looks like the parser needs some more work. It only accepts an object as the top level value.

File: lib/std/json.zig (hunk 2 above)
Comment (xackus) [reply lib/std/json.zig]: After checking, there are indeed some cases that are not caught. The tests only ensure that they don't crash the parser.

Comment (xackus) [reply lib/std/json.zig]: Why is `\u2abc` problematic?

Comment (xackus) [reply lib/std/json.zig]: I made the tests for json.StreamingParser run also for json.Parser.
That tests a lot of invalid stuff.

**Comment by xackus** [comment]
I added surrogate pair support.
I made the tests for json.StreamingParser run also for json.Parser and made them stricter to verify that correct errors are returned when given invalid surrogate pairs.
This uncovered some bugs, which I fixed (with the exception of an overflow in parseFloat, but that is not a parser bug).

I think this is ready to be merged.

File: lib/std/json.zig (hunk 2 above)
Comment (daurnimator) [reply lib/std/json.zig]: > Why is `\u2abc` problematic?

Doh, I meant to make something up like `\u2xyz`.

**Comment by andrewrk** [comment]
@xackus thanks for this pull request, and apologies for the delay in getting it merged.

File: lib/std/json.zig [hunk 3]
```diff
@@ -898,7 +898,13 @@ pub const TokenStream = struct {
             }
         }
 
-        if (self.parser.complete) {
+        // Without this a bare number fails, becasue the streaming parser doesn't know it ended
```
Comment (daurnimator) [inline lib/std/json.zig]: typo: 'becasue'

**Pull request merged**